src/modules/notary/index.ts

```diff
@@ -0,0 +1,168 @@
+type SignatureStr = string & { readonly [signature]: typeof signature };
+
+export type Validator<T> = (val: any) => Result<T>;
+export type Parser<T> = (str: JsonString<T>) => Result<T>;
+export type Formatter = (bytes: Buffer) => string;
+export type Signer = (utf8: Buffer) => Promise<Signature>;
+export type Verifier<T> = (utf8: Buffer, signature: Signature, payload: T) => Promise<boolean>;
+
+export const jsonStringify = <T>(value: T): JsonString<T> => JSON.stringify(value) as JsonString<T>;
+
+export const jsonParse =
+  <T>(validate: Validator<T>): Parser<T> =>
+  (string: JsonString<T>) => {
+    const ret = JSON.parse(string);
+
+    if (validate(ret)) return ret;
+
+    return undefined;
+  };
+
+export const ajvValidate =
+  <T>(validate: ValidateFunction<T>): Validator<T> =>
+  (val: any) => {
+    if (validate(val)) return { ok: val };
+
+    return { err: validate.errors?.join('; ') ?? 'Failed to validate object' };
+  };
+
+export const ajvParse =
+  <T>(parse: JTDParser<T>): Parser<T> =>
+  (val: string) => {
+    const ok = parse(val);
+    if (ok !== undefined) return { ok };
+
+    return { err: parse.message ?? 'Failed to parse JSON string' };
+  };
+
+export interface Notarized<T, S extends { [signature]: typeof signature } = Signature> {
+  payload: JsonString<T>;
+  signature: S;
+}
+export type NotarizedStr<T> = Notarized<T, SignatureStr>;
+
+const defaultFormat = (hash: Buffer) =>
+  `Your transaction fingerprint is: ${hash.toString('base64')}`;
+
+const createMessage = async (payloadStr: string, format: Formatter): Promise<Buffer> => {
+  const payloadBuf = Buffer.from(payloadStr, 'utf-8');
+  const payloadArr = new Uint8Array(
+    payloadBuf.buffer.slice(payloadBuf.byteOffset, payloadBuf.byteOffset + payloadBuf.byteLength)
+  );
+  const bytes = Buffer.from(await sha256(payloadArr, { outputFormat: 'buffer' }));
+
+  return Buffer.from(format(bytes), 'utf-8');
+};
+
+export const signPhantom =
+  (solana: Solana): Signer =>
+  async (utf8) =>
+    (await solana.signMessage(utf8, 'utf-8')).signature as Signature;
+
+export const signNacl =
+  (secret: Uint8Array): Signer =>
+  async (utf8) =>
+    Buffer.from(nacl.sign.detached(utf8, secret)) as Signature;
+
+export const verifyNacl =
+  (pubkey: Uint8Array): Verifier<unknown> =>
+  async (utf8, sig) =>
+    nacl.sign.detached.verify(utf8, sig, pubkey);
+
+export const verifyNaclSelfContained =
+  <T>(pubkey: (val: T) => Uint8Array): Verifier<T> =>
+  async (utf8, sig, payload) =>
+    nacl.sign.detached.verify(utf8, sig, pubkey(payload));
+
+export const notarize = async <T>(
+  payload: T,
+  sign: (utf8: Buffer) => Promise<Signature>,
+  options?: {
+    format?: (bytes: Buffer) => string;
+  }
+): Promise<Notarized<T>> => {
+  const payloadStr = jsonStringify(payload);
+  const msg = await createMessage(payloadStr, options?.format ?? defaultFormat);
+  const signature = await sign(msg);
+
+  return { payload: payloadStr, signature };
```

Comment:
Format payload in json. Create message for phantom to sign. Phantom signs message.